### PR TITLE
Enable i64 arguments with EM_ASM + WASM_BIGINT

### PR DESCRIPTION
**Pull request by sbc100**
Without WASM_BIGINT we assert in debug builds if i64
arguments are passed.

Partial fix for #15828

### CODE REVIEW

**Comment by sbc100** [comment]
One problem with this solution is that it won't work without WASM_BIGINT, but we can't inform the user until runtime when ASSERTIONS are enabled. 

It could be argued that we should instead just not fix this and instead ask folks to use EM_JS if they want to pass i64 arguments to JS.

**Comment by mmarczell-graphisoft** [comment]
@sbc100 Is there a reason you left out uint64_t (unsigned long long) support from this PR?

As a counterargument to your comment, EM_JS can also be considered incomplete wrt i64 support. First, this does not even assert when compiled without WASM_BIGINT:

```
EM_JS(void, foo, (int64_t i), {
  console.log(i);
})

int main() {
  foo(1);
}
```

Second, EM_ASM has a useful MAIN_THREAD_EM_ASM variant. With EM_JS we would have to use `emscripten_sync_run_in_main_runtime_thread`, but the EM_FUNC_SIG macros are also not complete with i64 variants.

Third, https://github.com/emscripten-core/emscripten/issues/13902 also affects EM_JS.

File: src/library.js [hunk 1]
```diff
@@ -3118,13 +3118,27 @@ LibraryManager.library = {
...
+      if (ch === 106) {
+        v = HEAP64[buf++ >> 1];
+      } else
+#endif
+      if (readAsmConstArgsDouble) {
+        v = HEAPF64[buf++ >> 1];
+      } else {
+        v = HEAP32[buf];
+      }
+      readAsmConstArgsArray.push(v);
```
Comment (juj) [inline src/library.js:3141]: How about:
```js

// Floats are always passed as doubles, and doubles and int64s take up 8 bytes (two 32-bit slots) in memory, align reads to these:
buf += (ch != 105) & buf;
readAsmConstArgsArray.push(
  ch == 105/*i*/ ? HEAP32[buf] :
#if WASM_BIGINT
   (ch == 106/*j*/ ? HEAP64 : HEAPF64)[buf++ >> 1]
#else
   HEAPF64[buf++ >> 1]
#endif
);
```
?

**Review by juj — APPROVED** [review]

File: src/library.js [hunk 2]
```diff
@@ -3118,13 +3118,27 @@ LibraryManager.library = {
...
+      assert(ch === 100/*'d'*/ || ch === 102/*'f'*/ || ch === 105 /*'i'*/ || ch === 106 /*'j'*/);
+#if !WASM_BIGINT
+      assert(ch !== 106/*'j'*/, "i64 arguments to ASM_JS function are not available without WASM_BIGINT");
+#endif
 #endif
       // A double takes two 32-bit slots, and must also be aligned - the backend
-      // will emit padding to avoid that.
-      var readAsmConstArgsDouble = ch < 105;
+      // will emit padding to avoid that.  Same with i64.
+      var readAsmConstArgsDouble = ch != 105
```
Comment (kripken) [inline src/library.js:3128]:
```suggestion
      var readAsmConstArgsDouble = ch != 105;
```
Maybe rename this to `read64Bits`?

File: src/library.js [hunk 3]
```diff
@@ -3118,13 +3118,27 @@ LibraryManager.library = {
...
+      var v;
       if (readAsmConstArgsDouble && (buf & 1)) buf++;
-      readAsmConstArgsArray.push(readAsmConstArgsDouble ? HEAPF64[buf++ >> 1] : HEAP32[buf]);
+#if WASM_BIGINT
+      if (ch === 106) {
+        v = HEAP64[buf++ >> 1];
+      } else
+#endif
+      if (readAsmConstArgsDouble) {
```
Comment (kripken) [inline src/library.js:3136]:
```suggestion
      if (ch == 105) {
```
This seems clearer as it avoids ambiguity around i64, I think.

File: system/include/emscripten/em_asm.h [hunk 4]
```diff
@@ -140,6 +141,7 @@ template<> struct __em_asm_sig<long> { static const char value = 'i'; };
 template<> struct __em_asm_sig<unsigned long> { static const char value = 'i'; };
 template<> struct __em_asm_sig<bool> { static const char value = 'i'; };
 template<> struct __em_asm_sig<wchar_t> { static const char value = 'i'; };
+template<> struct __em_asm_sig<long long> { static const char value = 'j'; };
```
Comment (mmarczell-graphisoft) [inline system/include/emscripten/em_asm.h:144]: Please include a specialization for `unsigned long long` too, if possible

**Comment by sbc100** [comment]
> 

Regarding uint64_t.. we don't currently support any unsigned values in EM_ASM.  For example if you pass UINT32_MAX today it will come out as `-1` on the JS side.   Adding support for unsigned values can be don't separately to this change as it seems orthogonal to just adding a new signed type.

File: system/include/emscripten/em_asm.h (hunk 4 above)
Comment (sbc100) [reply system/include/emscripten/em_asm.h:144]: Done.   However, just like with `uint32_t` these show up as signed on the JS side for now.   We can look at addressing that as a followup.

**Pull request merged**